### ISSUE TITLE
Get test case location in source via CLI

### ISSUE THREAD

**Issue by itavero** [cpputest/cpputest#967]
We are working on an initial version of a CppUTest Test Runner for Visual Studio.
So far, we are able to detect which tests are available from the compiled binary using the `-ln` flag.
We can run selected tests and process the JUnit XML output for displaying the results, errors and execution time.

The next step would be to be able to jump directly to the test case in the source code from the Test Explorer in Visual Studio.
I tried using the information from the JUnit XML output, so we could at least jump to failed assertions.
However, Visual Studio does not process any changes to the test case after discovery (I can only publish new test results, which basically is OK / Failed with an additional string... rather limited).
So, we need this information during initial test case discovery, for which we currently use the `-ln` flag.

As far as I can tell, there currently isn't a way to get this information via the command-line interface.
Perhaps I could decompile the binary or do some other smart analysis, but that all feels like "hacking" to me.

I'd suggest adding a more elaborate interface (perhaps also with XML or JSON output to allow for easier parsing) that exposes the following information (XML-ish example):
````xml
<testrunner>
  <group name="SomeTestGroup">
    <testcase name="FirstTest" source="C:\Path\To\Source.c" startsOnLine="38" />
    <testcase name="SomeIgnoredTest" source="C:\Path\To\Source.c" startsOnLine="52" ignored="true" />
  </group>
</testrunner>
````

If anyone has suggestions for additional information that would be nice to expose, I'd love to hear them.

PS: This is currently being developed for internal use at one of our clients. We are looking into open-sourcing the Visual Studio integration. :wink:

**Comment by arstrube** [comment]
I wonder whether this is really necessary? Have a look at this test runner for Eclipse. Perhaps it can give you some ideas:

git@ssh.github.com:tcmak/CppUTestEclipseJunoTestRunner.git

**Comment by arstrube** [comment]
As you will see from the sources, the usual way is to parse the console output. One has to look at the cdt testrunner in order to see the rest of the 'magic'. All I know is that the current Cli contains everything needed to make it work, in particular, the possibility to run any combination of tests.

I don't have any deep knowledge of this, but I believe it initially has no knowledge of which test groups and test cases are there. It derives this from the console output, as well as the source path to failed tests, so you can jump to the source of failed tests. I don't think it lets you jump to the source of successful tests (haven't used it for some time).

So it runs the entire test suite the first time, builds a graphical representation of test groups and tests, and lets you run tests selectively thereafter.

**Comment by itavero** [comment]
Please note that Eclipse and Visual Studio are very different products.
I'm not familiar with the Eclipse API for building a test runner, but a quick scan of the repository you mentioned tells me that it is rather different than the interface VS offers.
The API within Visual Studio is relatively new and unfortunately very poorly documented.
After more than 10 different approaches on updating the test location after the discovery phase, I'm absolutely sure I need to know it during discovery.

An alternative would be to somehow find the source files within the project and "guess" where the tests are using some regular expression magic.
But when we use macro's and such to generate similar tests (like I suggested in issue #939), this magic quickly becomes way less magical.
We used to discover tests the other way around: by scanning the source files using regular expressions to detect the `TEST` macro, instead of using the binary as input.
But, when you do that, it is hard to figure out how to actually run the tests (which is more important than knowing where they are) and again, introducing other macros broke the regex magic.

**Comment by itavero** [comment]
Also, I do not want to run all the test suites just for discovery.
I only want to run them when the user explicitly chooses to do so.
Especially on larger projects, running all tests every time a binary changes can consume a lot of precious time and that is not needed to just discover tests.

**Comment by arstrube** [comment]
Can't you run the entire test the first time (-v) and then parse the output? you can use the strings you find directly and feed them into the Cli on the next run.

As for source location, there will be one for each failed test. But I don't know the VS magic....

**Comment by arstrube** [comment]
Hmm. We use a script-based approach for heavy-duty. We need to collect coverage results as well, etc. etc.

The Eclispe test runner is just sugar for the developers, and most actually don't use it.

**Comment by arstrube** [comment]
What are you actually expecting that the test runner should do for you? I know only two use cases:

1. I want to run all tests, because that's the way to test my software.
2. I am busy working on a specific part and want to run one test (or a few) selectively.

Perhaps have a look at CppUTests way of using Cmake / Ctest. I don't like it much personally, but it seems to do a similar thing to what you wish to achieve.

**Comment by itavero** [comment]
I do so too, but the developers at this client have been using Visual Studio for years and are new to unit testing.
We want to make it as easy for them as we can, so we can motivate them to write tests, without asking them to use different scripts and command line magic and a lot of other "new stuff" (one step at a time). :wink:

**Comment by arstrube** [comment]
Okay.... that would be use case 2... I found the "Eclispe approach" (run the whole binary once, select specific tests therafter + be able to jump to failed tests) to work fine for that.

**Comment by itavero** [comment]
The test runner currently list all tests in the Test Explorer (which allows you to group them by status, or test group).
You can either choose "Run All", choose a single test to debug it, or select several test you wish to run (for example: all failed tests).
This all works and the results get shown in the Test Explorer to after the tests have been run.
So far so good and for now, this might be just enough.

However, just clicking the test in the Test Explorer instead of having to find the location of a test manually, would be a really nice feature.

**Comment by arstrube** [comment]
I checked the Eclipse runner. It does NOT have the ability to jump to any test.

However, if a test failed, it will display the failure message (which actually contains file/line info) and it will be able to jump to that location. I would imagine you could do the same in VS.

**Comment by arstrube** [comment]
The path/ file/ line info should be in the xml as well, for failures.

**Comment by itavero** [comment]
I know, but as I mentioned before, changing the `CodeFilePath` and `LineNumber` property of the `TestCase` instance does not update the user interface.
I also tried using the `SetPropertyValue` method, but this does not make the user interface show the location either nor will it allow you to jump to the location.
As said before, this information needs to be set during the Discovery phase (== before actually running the tests).
Just displaying the file in the error message or stacktrace will not make it "clickable" in VS.

**Comment by arstrube** [comment]
Is there an error message in the console window?

**Comment by arstrube** [comment]
Afaik, even when using no test runner of any sort, double-clicking on a failure msg in the console window should work.

Also, there is a setting in CppUTest, which makes the output Vs-friendly.

**Comment by itavero** [comment]
No, not in the console window, because it is not shown and the `-ojunit` flag is also passed to the executable.

**Comment by arstrube** [comment]
yes, -ojunit would prevent output to the console.

**Comment by arstrube** [comment]
Maybe you need something analogous to -oteamcity?

**Comment by itavero** [comment]
I know what `-ojunit` does and `-oteamcity` is not yet released and would still require me to actually run the tests.

**Comment by arstrube** [comment]
You can combine -v -junit to get both console output and junit output (don't know whether it is released, though). The user could then click the console window.

**Comment by itavero** [comment]
No, he can't. The console window is not visible. Tests are executed in a background process.

**Comment by arstrube** [comment]
Hmm.

**Comment by itavero** [comment]
And again, that would be a workaround, because it still would not allow me to click *any* of the tests and jump to it.

**Comment by arstrube** [comment]
Nobody can do that with any IDE :P. It hasn't killed us :D

**Comment by arstrube** [comment]
Suppose you found a way to do it via the Cli and suppose you contributed a PR to make it happen, then it would still take, errr, some time until you'd find that in a release?

**Comment by arstrube** [comment]
I made this test-runner-click-to-select thing available to a large group of developers. Nobody uses it anymore. They'd rather `#if 0` unwanted test cases temporarily :-/

**Comment by itavero** [comment]
I know and I'm not saying that this *must* be implemented right away. I'm just saying that providing this mechanism would enable people to implement such functionality, which would be very nice.

And the fact that releases take a long time might be annoying, but it can't be a reason not to implement something.

**Comment by itavero** [comment]
Also, I'm might eventually provide a PR for this myself, just putting it on here so everyone can see it and I don't forget.

**Comment by arstrube** [comment]
A PR always stand a fair chance. In Issue - if it's not a bug - may never be addressed.....

**Comment by itavero** [comment]
Again, this is not something that *must* be done. It would be *nice to have*. So that's okay.

**Comment by basvodde** [comment]
Interesting discussion.

Do you know if there is perhaps a standard format for this already?

**Comment by itavero** [comment]
I just found a newer version of the JUnit XML format:
https://github.com/jenkinsci/xunit-plugin/blob/master/src/main/resources/org/jenkinsci/plugins/xunit/types/model/xsd/junit-10.xsd

This includes a *file* and *line* option (also requested in #903 I believe).
It also describes a *skipped* tags.
If these would be included in the XML output, and I would have an option to skip/ignore all tests from the CLI, that would be a good solution too.

**Comment by itavero** [comment]
Or perhaps that could be the output when combining `-ln` and `-ojunit`...

**Comment by basvodde** [comment]
I'd probably not combine the parameters, but add -ljunit or so.

So, the output would be the latest junit output with all tests skipped?

**Comment by itavero** [comment]
I think that would be a decent approach. I've not yet come across another format that is widely used for retrieving this kind of information.

**Comment by itavero** [comment]
If I have time I'll dive into the Cpputest code tonight/this weekend to see if I can update the JUnit output and provide a mechanism to get these results. :wink:

**Comment by itavero** [comment]
By the way: is there a reason that XML output is generated per test group?
They could all be in a single file by putting the `<testsuite>` elements in an `<testsuites>` element.

**Comment by arstrube** [comment]
I don't know whther there is a reason - I think a single file output would be much nicer.

**Comment by itavero** [comment]
@arstrube So do I. Especially if I could set the file name (see #959) or perhaps do it like other tools do and just write it to standard out.

**Comment by itavero** [comment]
As you can see, I took the time to add the file and line attribute to the testcase element.
I figured that perhaps running the tests with  `-ojunit -xn` would be enough, but unfortunately filtering out a test isn't the same as ignoring one.

Also, the XSD I mentioned earlier also describes an `assertions` attribute for the `testcase` element.
While I'm at it, I might as well see if I can provide this too.
Does not seem like too much trouble to do so.

**Comment by itavero** [comment]
I've added the `assertions` attribute yesterday night.
Still need to add some unit tests for that.

Regarding the filtered tests: I'd propose adding `skipped` element for both the tests that have been ignored as well as those that have been filtered out.
The `skipped` element has a `type` attribute that can be used to distinguish the two, for example:
```xml
<?xml version="1.0" encoding="UTF-8" ?>
<testsuite tests="2">
  <testcase classname="TestGroup" name="IgnoredTest" assertions="0" file="source.cpp" line="20">
    <skipped type="ignore" />
  </testcase>
  <testcase classname="TestGroup" name="ExcludeFilterTest" assertions="0" file="source.cpp" line="30">
    <skipped type="filter" />
  </testcase>
</testsuite>
```

**Comment by itavero** [comment]
In `JUnitTestOutput::printCurrentTestEnded` I save the result of `result.getCheckCount()` to a field I've added to the `JUnitTestCaseResultNode`.
Unfortunately `getCheckCount()` seems to return the total number of checks performed so far, not just for the current test case.
Therefore, in the `while`-loop of `JUnitTestOutput::writeTestCases`, I'm storing the result to a local variable so I can subtract it in the next iteration.
This seems to be giving me the correct results.
However, the first item within a group might have a weird number, because I do not account for the checks in a previous test group.

I can probably make it work this way, but I was wondering if there is a method to just get the number of checks done within the test that has ended.

**Comment by itavero** [comment]
Managed to add to `assertions` attribute too.
Next up: investigate how to get the filtered tests to show up too (or make a flag to "ignore" all tests).
What would be the preferred solution?

**Comment by itavero** [comment]
@basvodde What's your opinion? Add an additional flag or refactor the code so that "filtered" tests also get through to the output classes?

Adding a flag sounds the "easiest" way to achieve this (and perhaps the intent is clear when using the flag, instead of calling cpputest with something like `-ojunit -xn`).
If the flag is the preferred solution, should we stick with `-ljunit` as you proposed earlier?

**Comment by itavero** [comment]
Been very busy because we are wrapping up the project at the current client.
I'll see if I can at least put in a PR for the additions I made to the JUnit output (which would resolve #120 and #903).
Then I'll make a separate PR for the additional flag to output all tests with this format later on.

**Comment by itavero** [comment]
As announced several weeks ago, I finally had the time to create a PR for the changes I made to the JUnit output (see #1010).

This does _not_ resolve _this_ issue, but I guess I'm halfway there.
